src/streaming/StreamProcessor.js

```diff
@@ -184,20 +234,74 @@ function StreamProcessor(config) {
         return representationController ? representationController.isUpdating() : false;
     }
 
+    function onStreamInitialized(e) {
+        if (!e.streamInfo || streamInfo.id !== e.streamInfo.id) return;
+
+        if (!streamInitialized) {
+            streamInitialized = true;
+            if (isDynamic) {
+                timelineConverter.setTimeSyncCompleted(true);
+                setLiveEdgeSeekTarget();
+            } else {
+                const seekTarget = playbackController.getStreamStartTime(false);
+                bufferController.setSeekStartTime(seekTarget);
+                scheduleController.setCurrentRepresentation(getRepresentationInfo());
+                scheduleController.setSeekTarget(seekTarget);
+            }
+        }
+
+        scheduleController.start();
+    }
+
     function onDataUpdateCompleted(e) {
-        if (e.sender.getType() !== getType() || e.sender.getStreamId() !== streamInfo.id || !e.error || e.error.code !== Errors.SEGMENTS_UPDATE_FAILED_ERROR_CODE) return;
+        if (e.sender.getType() !== getType() || e.sender.getStreamId() !== streamInfo.id) return;
 
-        addDVRMetric();
+        if (!e.error) {
+            scheduleController.setCurrentRepresentation(adapter.convertDataToRepresentationInfo(e.currentRepresentation));
+        } else if (e.error.code !== Errors.SEGMENTS_UPDATE_FAILED_ERROR_CODE) {
+            addDVRMetric();
+        }
+    }
+
+    function onQualityChanged(e) {
+        if (type !== e.mediaType || streamInfo.id !== e.streamInfo.id) return;
+        let representationInfo = getRepresentationInfo(e.newQuality);
+        scheduleController.setCurrentRepresentation(representationInfo);
+        dashMetrics.pushPlayListTraceMetrics(new Date(), PlayListTrace.REPRESENTATION_SWITCH_STOP_REASON);
+        dashMetrics.createPlaylistTraceMetrics(representationInfo.id, playbackController.getTime() * 1000, playbackController.getPlaybackRate());
     }
 
     function onBufferLevelUpdated(e) {
-        if (e.sender.getStreamProcessor() !== instance) return;
-        let manifest = manifestModel.getValue();
-        if (!manifest.doNotUpdateDVRWindowOnBufferUpdated) {
+        if (e.streamId !== streamInfo.id || e.mediaType !== type) return;
```

Comment:
We could add "dashMetrics.addBufferLevel()" at this point and remove dashMetrics from the BufferController.

Reply:
Yes, good point. I'll do it this way